# <center><span style="color:red;"> <b> Booking reviews </b></span></center>

<p align="center">
    <strong> </strong> <a href="https://www.kaggle.com/competitions/sf-booking"> Kaggle - Hotel Booking Link</a>
</p>

## <center><span style="text-decoration: underline;">Feature Engineering + Prediction of the Reviewer's Score </span></center>

<h4 style="margin-bottom: 0;">📝 Dataset Description: Hotel Reviews Dataset </h4>

- **`hotel_address`**: Address of the hotel.
- **`review_date`**: Date when the reviewer posted the corresponding review.
- **`average_score`**: Average score of the hotel, calculated based on the most recent comments over the past year.
- **`hotel_name`**: Name of the hotel.
- **`reviewer_nationality`**: Country of the reviewer.
- **`negative_review`**: Negative review provided by the reviewer.
- **`review_total_negative_word_counts`**: Total number of words in the negative review.
- **`positive_review`**: Positive review provided by the reviewer.
- **`review_total_positive_word_counts`**: Total number of words in the positive review.
- **`reviewer_score`**: Score assigned by the reviewer based on their experience.
- **`total_number_of_reviews_reviewer_has_given`**: Total number of reviews the reviewer has given in the past.
- **`total_number_of_reviews`**: Total number of valid reviews for the hotel.
- **`tags`**: Tags assigned by the reviewer to the hotel.
- **`days_since_review`**: Number of days between the review date and the dataset's last update.
- **`additional_number_of_scoring`**: Some guests provide a rating without leaving a review. This number indicates the count of valid ratings without associated reviews.
- **`lat`**: Latitude of the hotel.
- **`lng`**: Longitude of the hotel.

Source: ["Booking reviews" (kaggle.com)](https://www.kaggle.com/competitions/sf-booking) 

## <div style="text-align:center; border-radius:30px 30px; padding:7px; color:white; margin:0; font-size:110%; font-family:Pacifico; background-color:#3170af; overflow:hidden"><b> Table of Content </b></div>

1. [Import Libraries](#chapter1)
2. [Initialize Functions](#chapter2)
3. [Descriptive Statistics](#chapter3)
4. [Feature Engineering](#chapter4)
5. [Graphical + Statistical Analyses](#chapter5)    
    &emsp; 5.1 [Insurance Charges](#chapter5a)  
    &emsp; 5.2 [Insurance Charges and Gender](#chapter5b)  
    &emsp; 5.3 [Insurance Charges and Smoking Status](#chapter5c)  
    &emsp; 5.4 [Insurance Charges across Regions](#chapter5d)              
    &emsp; 5.5 [Insurance Charges across Age Groups](#chapter5e)   
    &emsp; 5.6 [Conclusions of EDA](#chapter5f)     
6. [Modeling](#chapter6)  
    &emsp; 6.1 [Linear Model](#chapter6a)     
    &emsp; 6.2 [Stochastic Gradient Descent](#chapter6b)    
    &emsp; 6.3  [Linear Model – Outliers adjustment](#chapter6c)   
    &emsp; 6.4  [Polynomials](#chapter6d)    
    &emsp; 6.5  [Polynomials with Regularization](#chapter6e)    
7.  [Conclusion](#chapter7)    

<a id="chapter1"></a>
## <div style="text-align:center; border-radius:30px 30px; padding:7px; color:white; margin:0; font-size:110%; font-family:Pacifico; background-color:#3170af; overflow:hidden"><b> ⚙️ Download Libraries </b></div>

In [ ]:
import pandas as pd 
import numpy as np

<a id="chapter3"></a>
## <div style="text-align:center; border-radius:30px 30px; padding:7px; color:white; margin:0; font-size:110%; font-family:Pacifico; background-color:#3170af; overflow:hidden"><b> ✨ Some preliminary descriptive analysis  </b></div>

In [6]:
# Upload Dataset
df = pd.read_csv('data/hotels.csv')
data_temp = df.copy()
display('Beginning of the dataset:')
display(data_temp.head(2))
display('End of the dataset:')
display(data_temp.head(2))
display("Number of variables:", data_temp.shape[1])
display('Number of observations:', data_temp.shape[0])

'Beginning of the dataset:'

hotel_address  \
0  Stratton Street Mayfair Westminster Borough Lo...   
1  130 134 Southampton Row Camden London WC1B 5AF...   

   additional_number_of_scoring review_date  average_score  \
0                           581   2/19/2016            8.4   
1                           299   1/12/2017            8.3   

                        hotel_name reviewer_nationality  negative_review  \
0               The May Fair Hotel      United Kingdom          Leaving    
1  Mercure London Bloomsbury Hotel      United Kingdom    poor breakfast   

   review_total_negative_word_counts  total_number_of_reviews  \
0                                  3                     1994   
1                                  3                     1361   

       positive_review  review_total_positive_word_counts  \
0   Staff were amazing                                  4   
1             location                                  2   

   total_number_of_reviews_reviewer_has_given  reviewer_score  \
0                                           7            10.0   
1                                          14             6.3   

                                                tags days_since_review  \
0  [' Leisure trip ', ' Couple ', ' Studio Suite ...           531 day   
1  [' Business trip ', ' Couple ', ' Standard Dou...           203 day   

         lat       lng  
0  51.507894 -0.143671  
1  51.521009 -0.123097

'End of the dataset:'

hotel_address  \
0  Stratton Street Mayfair Westminster Borough Lo...   
1  130 134 Southampton Row Camden London WC1B 5AF...   

   additional_number_of_scoring review_date  average_score  \
0                           581   2/19/2016            8.4   
1                           299   1/12/2017            8.3   

                        hotel_name reviewer_nationality  negative_review  \
0               The May Fair Hotel      United Kingdom          Leaving    
1  Mercure London Bloomsbury Hotel      United Kingdom    poor breakfast   

   review_total_negative_word_counts  total_number_of_reviews  \
0                                  3                     1994   
1                                  3                     1361   

       positive_review  review_total_positive_word_counts  \
0   Staff were amazing                                  4   
1             location                                  2   

   total_number_of_reviews_reviewer_has_given  reviewer_score  \
0                                           7            10.0   
1                                          14             6.3   

                                                tags days_since_review  \
0  [' Leisure trip ', ' Couple ', ' Studio Suite ...           531 day   
1  [' Business trip ', ' Couple ', ' Standard Dou...           203 day   

         lat       lng  
0  51.507894 -0.143671  
1  51.521009 -0.123097

'Number of variables:'

17

'Number of observations:'

386803

In [7]:
data_temp.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 386803 entries, 0 to 386802
Data columns (total 17 columns):
 #   Column                                      Non-Null Count   Dtype  
---  ------                                      --------------   -----  
 0   hotel_address                               386803 non-null  object 
 1   additional_number_of_scoring                386803 non-null  int64  
 2   review_date                                 386803 non-null  object 
 3   average_score                               386803 non-null  float64
 4   hotel_name                                  386803 non-null  object 
 5   reviewer_nationality                        386803 non-null  object 
 6   negative_review                             386803 non-null  object 
 7   review_total_negative_word_counts           386803 non-null  int64  
 8   total_number_of_reviews                     386803 non-null  int64  
 9   positive_review                             386803 non-null  object 
 

In [ ]:
data_temp.describe() 

additional_number_of_scoring  average_score  \
count                 386803.000000  386803.000000   
mean                     498.246536       8.397231   
std                      500.258012       0.547881   
min                        1.000000       5.200000   
25%                      169.000000       8.100000   
50%                      342.000000       8.400000   
75%                      660.000000       8.800000   
max                     2682.000000       9.800000   

       review_total_negative_word_counts  total_number_of_reviews  \
count                      386803.000000            386803.000000   
mean                           18.538988              2743.992042   
std                            29.703369              2316.457018   
min                             0.000000                43.000000   
25%                             2.000000              1161.000000   
50%                             9.000000              2134.000000   
75%                            23.000000              3613.000000   
max                           408.000000             16670.000000   

       review_total_positive_word_counts  \
count                      386803.000000   
mean                           17.776985   
std                            21.726141   
min                             0.000000   
25%                             5.000000   
50%                            11.000000   
75%                            22.000000   
max                           395.000000   

       total_number_of_reviews_reviewer_has_given  reviewer_score  \
count                                386803.00000   386803.000000   
mean                                      7.17725        8.396906   
std                                      11.05442        1.636090   
min                                       1.00000        2.500000   
25%                                       1.00000        7.500000   
50%                                       3.00000        8.800000   
75%                                       8.00000        9.600000   
max                                     355.00000       10.000000   

                 lat            lng  
count  384355.000000  384355.000000  
mean       49.443522       2.823402  
std         3.466936       4.579043  
min        41.328376      -0.369758  
25%        48.214662      -0.143649  
50%        51.499981      -0.000250  
75%        51.516288       4.834443  
max        52.400181      16.429233

In [9]:
data_temp.describe(include='object')

hotel_address review_date  \
count                                              386803      386803   
unique                                               1493         731   
top     163 Marsh Wall Docklands Tower Hamlets London ...    8/2/2017   
freq                                                 3587        1911   

                                        hotel_name reviewer_nationality  \
count                                       386803               386803   
unique                                        1492                  225   
top     Britannia International Hotel Canary Wharf      United Kingdom    
freq                                          3587               184033   

       negative_review positive_review  \
count           386803          386803   
unique          248828          311737   
top        No Negative     No Positive   
freq             95907           26885   

                                                     tags days_since_review  
count                                              386803            386803  
unique                                              47135               731  
top     [' Leisure trip ', ' Couple ', ' Double Room '...            1 days  
freq                                                 3853              1911

In [10]:
if data_temp.isin([np.inf, -np.inf]).any().any():
    print("Infinite values found!")
else:
    print("No infinite values found!")    

# Check for missing values
if data_temp.isnull().any().any():
    print("Missing values found!")
else:
    print("No missing values found!")

No infinite values found!
Missing values found!


In [ ]:
data_null = (data_temp.isnull().mean()*100).round(3).to_frame()
data_null.columns = ['Missing %']
data_null = data_null.rename_axis('Variables')
 
display(data_null)

Missing %
Variables                                            
hotel_address                                   0.000
additional_number_of_scoring                    0.000
review_date                                     0.000
average_score                                   0.000
hotel_name                                      0.000
reviewer_nationality                            0.000
negative_review                                 0.000
review_total_negative_word_counts               0.000
total_number_of_reviews                         0.000
positive_review                                 0.000
review_total_positive_word_counts               0.000
total_number_of_reviews_reviewer_has_given      0.000
reviewer_score                                  0.000
tags                                            0.000
days_since_review                               0.000
lat                                             0.633
lng                                             0.633

<div style="padding: 20px; border: 2px solid #c77220; border-radius: 5px; box-shadow: 0 2px 4px rgba(0, 0, 0, 0.1); max-width: 100%; margin: 0 auto;">
    <ul style="font-size: 18px; font-family: 'Arial', sans-serif; line-height: 1.5em; word-wrap: break-word; overflow-wrap: break-word;">
    <h4 style="margin-bottom: 0;">💡 Interpretation:</h4>
        <li>  Two variables (<code>lat</code> and <code>lng</code>) have missing observations of around 0.633%.  </li>
    </ul>
</div>